imap_processing/spice/geometry.py

```diff
@@ -1 +1,128 @@
+
+from collections.abc import Iterable
+from enum import Enum
+from typing import NamedTuple, Optional, Union
+
+import numpy as np
+import spiceypy as spice
+
+from imap_processing.spice.kernels import ensure_spice
+
+
+class SpiceId(NamedTuple):
+    """Class that represents a unique identifier in the NAIF SPICE library."""
+
+    strid: str
+    numid: int
+
+
+class SpiceBody(Enum):
+    """Enum containing SPICE IDs for bodies that we use."""
+
+    # A subset of IMAP Specific bodies as defined in imap_wkcp.tf
+    IMAP = SpiceId("IMAP", -43)
+    IMAP_SPACECRAFT = SpiceId("IMAP_SPACECRAFT", -43000)
+    # IMAP Pointing Frame (Despun) as defined in iamp_science_0001.tf
+    IMAP_DPS = SpiceId("IMAP_DPS", -43901)
+    # Standard NAIF bodies
+    SSB = SpiceId("SOLAR_SYSTEM_BARYCENTER", 0)
+    SUN = SpiceId("SUN", 10)
+    EARTH = SpiceId("EARTH", 399)
+
+
+class SpiceFrame(Enum):
+    """Enum containing SPICE IDs for reference frames, defined in imap_wkcp.tf."""
+
+    # Standard SPICE Frames
+    J2000 = SpiceId("J2000", 1)
+    ECLIPJ2000 = SpiceId("ECLIPJ2000", 17)
+    # IMAP specific
+    IMAP_SPACECRAFT = SpiceId("IMAP_SPACECRAFT", -43000)
+    IMAP_LO_BASE = SpiceId("IMAP_LO_BASE", -43100)
+    IMAP_LO_STAR_SENSOR = SpiceId("IMAP_LO_STAR_SENSOR", -43103)
+    IMAP_LO = SpiceId("IMAP_LO", -43105)
+    IMAP_HI_45 = SpiceId("IMAP_HI_45", -43150)
+    IMAP_HI_90 = SpiceId("IMAP_HI_90", -43160)
+    IMAP_ULTRA_45 = SpiceId("IMAP_ULTRA_45", -43200)
+    IMAP_ULTRA_90 = SpiceId("IMAP_ULTRA_90", -43210)
+    IMAP_MAG = SpiceId("IMAP_MAG", -43250)
+    IMAP_SWE = SpiceId("IMAP_SWE", -43300)
+    IMAP_SWAPI = SpiceId("IMAP_SWAPI", -43350)
+    IMAP_CODICE = SpiceId("IMAP_CODICE", -43400)
+    IMAP_HIT = SpiceId("IMAP_HIT", -43500)
+    IMAP_IDEX = SpiceId("IMAP_IDEX", -43700)
+    IMAP_GLOWS = SpiceId("IMAP_GLOWS", -43750)
+
+
+def imap_state(
+    et: Union[np.ndarray, float],
+    ref_frame: Optional[SpiceFrame] = None,
+    observer: Optional[SpiceBody] = None,
+) -> Union[np.ndarray, Iterable[np.ndarray]]:
+    """
+    Get the state (position and velocity) of the IMPA spacecraft.
+
+    By default, the state is returned in the ECLIPJ2000 frame as observed by the Sun.
+
+    Parameters
+    ----------
+    et : np.ndarray or float
+        Epoch time(s) [J2000 seconds] to get the IMAP state for.
+    ref_frame : SpiceFrame, optional
+        Reference frame which the IMAP state is expressed in.
+    observer : SpiceBody, optional
+        Observing body.
+
+    Returns
+    -------
+    state : np.ndarray or Iterable[np.ndarray]
+     The Cartesian state vector representing the position and velocity of the
+     IMAP spacecraft.
+    """
+    if ref_frame is None:
+        ref_frame = SpiceFrame.ECLIPJ2000
+    if observer is None:
+        observer = SpiceBody.SUN
+    state, light_time = ensured_spkezr(
+        SpiceBody.IMAP.name, et, ref_frame.name, "NONE", observer.name
+    )
+    return state
+
+
+def ensured_spkezr(
```

Comment:
How would you feel about making these more human readable? I think this is actually one of the things I don't like about `spiceypy`, where they just reiterated the low-level spice function names. So then I need to know SPICE to be able to use it.

If you make this: `calc_spacecarft_state()` I would find that much easier to parse what is going on. Then only you and other SPICE experts need to know this is ensured and using the underlying spkezr function under the hood.

Also, how many of these arguments will we actually want to change? Maybe limit the number to the relevant ones we'll modify and chosen defaults for the others.

Reply:
Perhaps I just need to make this a private function? The intent here is only to wrap the spicey function with `ensure_spice`. The general paradigm to follow when using `ensure_spice` is to wrap the lowest level functions you can. That reduces ensure spice recursion and lowers the processing done on the wrapped function retries.
I provided the above function `imap_state()` that does what your `calc_spacecraft_state()` function would do and is intended to be used by the team rather than this function.

Reply:
Ahh, I see, I don't know how I missed that you actually were making the nicer name. Thank you!!

This is where I think the global state and loading of kernels would be beneficial, and then you could wrap your original function with `@ensure_spice` rather than the low-level versions as a two-step process. No need to do the global thing here, I'm going to give that some thought and try to propose something.

> Perhaps I just need to make this a private function?
Yes, I think that might be good. The other possible thing to do is just inline the ensure_spice into your original function? It is only a few lines.